packages/cw-derive/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+use convert_case::{Case, Casing};
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use syn::fold::Fold;
+use syn::parse::{Parse, Parser};
+use syn::spanned::Spanned;
+use syn::{parse_macro_input, Error, FnArg, Ident, ItemTrait, Pat, TraitItem, TraitItemMethod};
+
+mod parser;
+mod strip_input;
+
+use strip_input::StripInput;
+
+/// Macro generating messages from contract trait.
+///
+/// ## Example usage
+/// ```ignore
+/// # use cosmwasm_std::Response;
+///
+/// # struct Ctx;
+/// # struct Error;
+///
+/// # #[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq, schemars::JsonSchema)]
+/// # struct Member;
+///
+/// #[cw_derive::interface(module=msg, exec=Execute, query=Query)]
+/// trait Cw4 {
+///     #[msg(exec)]
+///     fn update_admin(&self, ctx: Ctx, admin: Option<String>) -> Result<Response, Error>;
+///
+///     #[msg(exec)]
+///     fn update_members(&self, ctx: Ctx, remove: Vec<String>, add: Vec<Member>)
```

Comment:
I don't understand - if it is equivalent, then how they get different code generated? Schema generator claims it generates same schemas for equivalend serde serializers... The problem is that extracting this to structures may lead to names collisions easly, especially for more complicated structures. It is partially solved by keeping those in separate module (opt-in), but what if both query and execute would have same variant? Two same structures would be generated... You can also try to add suffixes on such case (like numbers) but will it still be easier to use? It is possible, but not sure how much would it improve at this level.

Reply:
NP. I don't see the name collision issue, but no need to do it.

The generated types have the same shape, but those intermediate type definitions (one per variant) can help in other languages. We can also work on the codegen side in other languages here.

That said, we can always change the auto-gen later and it would be functionally equivalent